tutorial/source/tensor_shapes.ipynb

```diff
@@ -0,0 +1,555 @@
+    "    with x_axis:\n",
+    "        x = pyro.sample(\"x\", Normal(0, 1).reshape([3, 1]))\n",
+    "    with y_axis:\n",
+    "        y = pyro.sample(\"y\", Normal(0, 1).reshape([2, 1, 1]))\n",
+    "    with x_axis, y_axis:\n",
+    "        xy = pyro.sample(\"xy\", Normal(0, 1).reshape([2, 3, 1]))\n",
+    "        z = pyro.sample(\"z\", Normal(0, 1).reshape([2, 3, 1, 5], 1))\n",
+    "    assert x.shape == (3, 1)        # batch_shape == (3,1)     event_shape == ()\n",
+    "    assert y.shape == (2, 1, 1)     # batch_shape == (2,1,1)   event_shape == ()\n",
+    "    assert xy.shape == (2, 3, 1)    # batch_shape == (2,3,1)   event_shape == ()\n",
+    "    assert z.shape == (2, 3, 1, 5)  # batch_shape == (2,3,1)   event_shape == (5,)\n",
+    "    \n",
+    "test_model(model1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It is helpful to visualize the `.shape`s of each sample site by aligning them at the boundary between `batch_shape` and `event_shape`: dimensions to the right will be summed out in `.log_prob()` and dimensions to the left will remain. \n",
+    "```\n",
+    "batch dims | event dims\n",
+    "-----------+-----------\n",
+    "           |        a = sample(\"a\", Normal(0, 1))\n",
+    "           |2       b = sample(\"b\", Normal(zeros(2), 1)\n",
+    "           |                        .reshape(extra_event_dims=1)\n",
+    "           |        with iarange(\"c\", 2):\n",
+    "          2|            c = sample(\"c\", Normal(zeros(2), 1))\n",
+    "           |        with iarange(\"d\", 3):\n",
+    "          3|4 5         d = sample(\"d\", Normal(zeros(3,4,5), 1)\n",
+    "           |                       .reshape(extra_event_dims=2)\n",
+    "           |\n",
+    "           |        x_axis = iarange(\"x\", 3, dim=-2)\n",
+    "           |        y_axis = iarange(\"y\", 2, dim=-3)\n",
+    "           |        with x_axis:\n",
+    "        3 1|            x = sample(\"x\", Normal(0, 1).reshape([3, 1]))\n",
+    "           |        with y_axis:\n",
+    "      2 1 1|            y = sample(\"y\", Normal(0, 1).reshape([2, 1, 1]))\n",
+    "           |        with x_axis, y_axis:\n",
+    "      2 3 1|            xy = sample(\"xy\", Normal(0, 1).reshape([2, 3, 1]))\n",
+    "      2 3 1|5           z = sample(\"z\", Normal(0, 1).reshape([2, 3, 1, 5], 1))\n",
+    "```\n",
+    "As an exercise, try to tabulate the shapes of sample sites in one of your own programs."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Subsampling tensors inside an `iarange`    <a class=\"anchor\" id=\"subsampling\"></a>\n",
+    "\n",
+    "One of the main uses of [iarange](http://docs.pyro.ai/en/dev/primitives.html#pyro.__init__.iarange) is to subsample data. This is possible within an `iarange` because data are independent, so the expected value of the loss on half the data should be half the expected value of the loss on the full data.\n",
```

Comment:
"so the expected value of the loss on half the data " ==>
so the expected value of the loss on, say, half the data